src/common/debug.c

```diff
@@ -238,12 +237,19 @@ daos_key2str(daos_key_t *key)
 				break;
 			}
 		}
-		if (can_print) {
+
+		if (can_print)
 			strncpy(buf, key->iov_buf, len);
-			buf[len] = 0;
-		} else {
+		else
 			strcpy(buf, "????");
+
+		len = min(len, strlen(buf));
```

Comment:
"buf" may be not "\0" terminated or terminated at an unexpected position, that will cause strlen() gives a wrong value.

Reply:
If strlen(buf) < the buffer real size, then the following "Uint64:xxx" will be appended at the end of string anyway.

otherwise, len will keep the same?

Reply:
The "strlen(buf)" may be smaller than "len" but larger than its real length, is it expected?

Reply:
len is the real size of buffer. min(len, strlen(buf)) should not be larger than the real size.

Reply:
For example, key->iov_len = 4, DF_KEY_STR_SIZE = 64, so "len" is 4 originally, but the key string may be not "\0" terminated. So when copy the key string into the "buf", the buf[0-3] is the key string, but "buf[4]" is not "\0", instead, it is former left things, then strlen(buf) may return 5/6/... or larger, which is not the real length of key string. That is not expected behavior. Such out of boundary strlen() may not cause segment fault, but looks strange.

Reply:
even strlen(buf) is larger than 4, but min(len, strlen(buf)) will be 4 in this case? or I miss sth?

Reply:
The final result is not wrong. But I am afraid that strlen() agains non-null terminated buffer may cause some potential trouble. If unfortunately out of boundary to some inaccessible area, then may be segment fault? I am not sure. So please confirm. Thanks!

Reply:
min(len, strlen()) will make sure it will not out of boundary at all. IMHO.

Reply:
I mean before min(), just inside strlen().

Reply:
hmm, I do not think strlen itself will cause any issue here, even the real size might be larger than buffer itself.

Reply:
It may be better regardless to use strnlen here if for no other reason than coverity hates strlen.   But it can also prevent any possibility of long buffer overrrun

Reply:
This is wrong. If key->iov_len is more than 64 it will copy only first 64 chars without trailing \0. So, strlen(buf) will count 64 plus all chars after the buf till \0. If you would like to insert \0 anyway please use sprintf() for this.

Reply:
>This is wrong. If key->iov_len is more than 64 it will copy only first 64 chars without trailing \0.

 int len = min(key->iov_len, DF_KEY_STR_SIZE - 1);  will make sure strncpy(buf, key->iov_buf, len); will always 1 byte for "/0".

Reply:
Yes, "len = min(key->iov_len, DF_KEY_STR_SIZE - 1);" is right. But not "len = min(len, strlen(buf));"

Reply:
Ok. got it after discussion. Very confusing code. I'd like to avoid such style or add explicit comments.

Reply:
Anyway it's better to use snprintf() here for all cases.